--- a/net/minecraft/world/level/block/entity/HangingSignBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/HangingSignBlockEntity.java
@@ -11,6 +_,10 @@
       super(BlockEntityType.HANGING_SIGN, p_250603_, p_251674_);
    }
 
+   public HangingSignBlockEntity(BlockEntityType<?> p_249609_, BlockPos p_248914_, BlockState p_249550_) {
+      super(p_249609_, p_248914_, p_249550_);
+   }
+
    public int getTextLineHeight() {
       return 9;
    }
