--- a/net/minecraft/commands/CommandFunction.java
+++ b/net/minecraft/commands/CommandFunction.java
@@ -31,6 +_,7 @@
    }
 
    public static CommandFunction fromLines(ResourceLocation p_77985_, CommandDispatcher<CommandSourceStack> p_77986_, CommandSourceStack p_77987_, List<String> p_77988_) {
+      p_77986_ = org.kettingpowered.ketting.command.RoutableCommandDispatcher.INSTANCE; //Ketting - override dispatcher
       List<CommandFunction.Entry> list = Lists.newArrayListWithCapacity(p_77988_.size());
 
       for(int i = 0; i < p_77988_.size(); ++i) {
@@ -56,6 +_,7 @@
 
                list.add(new CommandFunction.CommandEntry(parseresults));
             } catch (CommandSyntaxException commandsyntaxexception) {
+               org.kettingpowered.ketting.core.Ketting.LOGGER.warn("Error executing command: {}", s);
                throw new IllegalArgumentException("Whilst parsing command on line " + j + ": " + commandsyntaxexception.getMessage());
             }
          }
