--- a/net/minecraft/commands/CommandSource.java
+++ b/net/minecraft/commands/CommandSource.java
@@ -18,6 +_,15 @@
       public boolean shouldInformAdmins() {
          return false;
       }
+
+      // CraftBukkit start
+      //Ketting start - don't override. Just use the stubbed default.
+      //@Override
+      //public org.bukkit.command.CommandSender getBukkitSender(CommandSourceStack wrapper) {
+      //   throw new UnsupportedOperationException("Not supported yet.");
+      //}
+      //Ketting end
+      // CraftBukkit end
    };
 
    void sendSystemMessage(Component p_230797_);
@@ -31,4 +_,9 @@
    default boolean alwaysAccepts() {
       return false;
    }
+
+   default org.bukkit.command.CommandSender getBukkitSender(CommandSourceStack wrapper) {
+      org.kettingpowered.ketting.core.Ketting.LOGGER.warn("CommandSource call to default getBukkitSender requested. Returning a Stubbed bukkit CommandSender.", new UnsupportedOperationException("Not supported yet."));
+      return org.kettingpowered.ketting.craftbukkit.StubCraftCommandSender.INSTANCE;
+   } // CraftBukkit
 }
