--- a/net/minecraft/commands/CommandSource.java
+++ b/net/minecraft/commands/CommandSource.java
@@ -18,6 +_,16 @@
       public boolean shouldInformAdmins() {
          return false;
       }
+
+      // CraftBukkit start
+      @Override
+      public org.bukkit.command.CommandSender getBukkitSender(CommandSourceStack wrapper) {
+         //Ketting start - don't throw, don't warn. Just use the stubbed default.
+         //throw new UnsupportedOperationException("Not supported yet.");
+         return org.kettingpowered.ketting.craftbukkit.StubCraftCommandSender.INSTANCE;
+         //Ketting end
+      }
+      // CraftBukkit end
    };
 
    void sendSystemMessage(Component p_230797_);
@@ -31,4 +_,9 @@
    default boolean alwaysAccepts() {
       return false;
    }
+
+   default org.bukkit.command.CommandSender getBukkitSender(CommandSourceStack wrapper) {
+      org.kettingpowered.ketting.core.Ketting.LOGGER.warn("CommandSource call to default getBukkitSender requested. Returning a Stubbed bukkit CommandSender, to avoid crashing.", new UnsupportedOperationException("Not supported yet."));
+      return org.kettingpowered.ketting.craftbukkit.StubCraftCommandSender.INSTANCE;
+   } // CraftBukkit
 }
