--- a/net/minecraft/world/entity/vehicle/ContainerEntity.java
+++ b/net/minecraft/world/entity/vehicle/ContainerEntity.java
@@ -103,6 +_,9 @@
 
          this.setLootTable((ResourceLocation)null);
          LootParams.Builder lootparams$builder = (new LootParams.Builder((ServerLevel)this.level())).withParameter(LootContextParams.ORIGIN, this.position());
+         // Forge: set the chest to killer_entity for loot context.
+         if (this instanceof AbstractMinecartContainer entityContainer)
+            lootparams$builder.withParameter(LootContextParams.KILLER_ENTITY, entityContainer);
          if (p_219950_ != null) {
             lootparams$builder.withLuck(p_219950_.getLuck()).withParameter(LootContextParams.THIS_ENTITY, p_219950_);
          }
@@ -171,6 +_,6 @@
    }
 
    default boolean isChestVehicleStillValid(Player p_219955_) {
-      return !this.isRemoved() && this.position().closerThan(p_219955_.position(), 8.0D);
+      return !this.isRemoved() && this.position().closerThan(p_219955_.position(), p_219955_.getAttributeValue(net.minecraftforge.common.ForgeMod.ENTITY_REACH.get()) + 5.0); // Vanilla uses 8 default reach is 3.0 so add 5.0 padding
    }
 }
